docs/pages/parallel-async.md

```diff
@@ -0,0 +1,35 @@
+---
+title: Parallel and asynchronous processing
+layout: default
+---
+
+Python has a good ecosystem of libraries for multiprocessing (threads and GPU
+parallelisation), as well as asynchronous processing. Here, we list those that
+we've found to be useful, particularly for research applications and previous
+ARC projects.
```

Comment:
```suggestion
Python has a good ecosystem of libraries for parallelising the processing of tasks,
as well as asynchronous processing. While Python has built-in support for multithreading via 
[the `threading` module in the standard library](https://docs.python.org/3/library/threading.html)
the _global interpreter lock_ (GIL) [currently](https://peps.python.org/pep-0703/) 
means that only one thread can be execute Python code at a time. Parallelisation in Python 
therefore typically either is _process-based_ with code parallelised across multiple Python 
processes each with their own interpreter (and GIL) or makes use of tools which run the 
tasks to be parallelised outside of the Python interpreter, using for example Python 
wrappers around external code which uses thread-based parallelism. For the latter case,
some tools use code generation and just-in-time compilation to allow writing code in a 
Python like syntax while still supporting native thread based parallelism. Increasingly there
is also support for parallelising operations on accelerator devices such as graphics 
processing units.
 
Here, we list Python tools that offer support for parallelisation that
we've found to be useful, particularly for research applications and previous
ARC projects.
```

I think it might be worth adding some comment about `threading` / support for thread-based parallelism in Python as it might otherwise seem odd why we don't recommend using the `threading` module for someone who is not familiar with GIL (which is technically a CPython implementation detail so some of my edits above should say CPython instead of Python but I assumed we don't want to go down this whole different rabbit hole if trying to keep things relatively approachable for beginners). I suspect my edits above probably have too much technical detail and might be better in a comment after the actual recommendations, but I think it is worth mentioning the two main approaces to parallelism and why typically we use process based parallelism rather than threads (which would be the default in most other languages).

Reply:
Should we put threading in the table as :red_circle: ?

Reply:
Yeah that could work as a more succinct way of getting across the same point

Reply:
[1f8f261](https://github.com/UCL-ARC/python-tooling/pull/323/commits/1f8f26132ef85eee197cbc0fa646cb8a91d329b2)